tx/textual/valuerenderer: formatDecimal accepts non-ASCII and non-numeric/formatting characters like emojis and letters

<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺ 
v                            ✰  Thanks for opening an issue! ✰    
v    Before smashing the submit button please review the template.
v    Please also ensure that this is not a duplicate issue :)  
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

<!--
IMPORTANT: Prior to opening a bug report, check if it affects one of the core modules
and if its elegible for a bug bounty on `SECURITY.md`. Bugs that are not submitted
through the appropriate channels won't receive any bounty.
 -->

## Summary of Bug
Noticed in an audit of this package: by definition of a decimal number, I'd presume that formatDecimal is supposed to accept only sign values, numeric values and decimal points; but even more the docs for SIGNMODE_TEXTUAL say that it MUST only accept ASCII characters per https://github.com/cosmos/cosmos-sdk/blob/main/docs/architecture/adr-050-sign-mode-textual.md#2-only-ascii-32-127-characters-allowed
![Screen Shot 2022-08-03 at 5 24 55 PM](https://user-images.githubusercontent.com/4898263/182737445-71ebe215-5073-4dc1-9310-2ceec4eb0b4c.png)

and this is a demo
```go
package valuerenderer
  
import "testing"
  
func TestFormatxDecimal(t *testing.T) {
        s, err := formatDecimal("-😎.😎")
        if err == nil {
                t.Fatalf("Expected an error yet Passed with: %q\n", s)
        }
}
```

which produces
```shell
$ go test -run=txDecimal
--- FAIL: TestFormatxDecimal (0.00s)
    dec_test.go:8: Expected an error yet Passed with: "-\xf0'\x9f\x98\x8e.😎"
FAIL
exit status 1
FAIL	cosmossdk.io/tx/textual/valuerenderer	0.187s
```

and even passing in like "abcdef.fghghghg" passes :-(
<!-- Concisely describe the issue -->

## Fix
The proper way to fix this is to:
* trim leading and trailing whitespaces
* do a character set check with perhaps a compiled regex and then reject anything that isn't in that character set

## Version
ccc8003a8096a1eaa5f3597f011b038efd05e375
<!-- git commit hash or release version -->

Kindly cc-ing @amaurym @marbar3778 @elias-orijtech @willpoint @kirbyquerby

--- a/tx/textual/valuerenderer/bench_test.go
+++ b/tx/textual/valuerenderer/bench_test.go
@@ -9,15 +9,15 @@ import (
 )
 
 var intValues = []protoreflect.Value{
-	protoreflect.ValueOfString("10.00"),
-	protoreflect.ValueOfString("999.00"),
-	protoreflect.ValueOfString("999.9999"),
-	protoreflect.ValueOfString("99999999.9999"),
+	protoreflect.ValueOfString("1000"),
+	protoreflect.ValueOfString("99900"),
+	protoreflect.ValueOfString("9999999"),
+	protoreflect.ValueOfString("999999999999"),
 	protoreflect.ValueOfString("9999999999999999999"),
-	protoreflect.ValueOfString("1000000000000000000000000000000000000000000000000000000.00"),
-	protoreflect.ValueOfString("77777777777.777777777777777777777700"),
-	protoreflect.ValueOfString("-77777777777.777777777777777777777700"),
-	protoreflect.ValueOfString("777777777777777777777777.77777777700"),
+	protoreflect.ValueOfString("100000000000000000000000000000000000000000000000000000000"),
+	protoreflect.ValueOfString("77777777777777777777777777777777700"),
+	protoreflect.ValueOfString("-77777777777777777777777777777777700"),
+	protoreflect.ValueOfString("77777777777777777777777777777777700"),
 }
 
 func BenchmarkIntValueRendererFormat(b *testing.B) {
@@ -30,6 +30,35 @@ func BenchmarkIntValueRendererFormat(b *testing.B) {
 	for i := 0; i < b.N; i++ {
 		for _, value := range intValues {
 			if err := ivr.Format(ctx, value, buf); err != nil {
+				b.Fatal(err)
+			}
+		}
+		buf.Reset()
+	}
+}
+
+var decimalValues = []protoreflect.Value{
+	protoreflect.ValueOfString("10.00"),
+	protoreflect.ValueOfString("999.00"),
+	protoreflect.ValueOfString("999.9999"),
+	protoreflect.ValueOfString("99999999.9999"),
+	protoreflect.ValueOfString("9999999999999999999"),
+	protoreflect.ValueOfString("1000000000000000000000000000000000000000000000000000000.00"),
+	protoreflect.ValueOfString("77777777777.777777777777777777777700"),
+	protoreflect.ValueOfString("-77777777777.777777777777777777777700"),
+	protoreflect.ValueOfString("777777777777777777777777.77777777700"),
+}
+
+func BenchmarkDecimalValueRendererFormat(b *testing.B) {
+	ctx := context.Background()
+	dvr := new(decValueRenderer)
+	buf := new(bytes.Buffer)
+	b.ResetTimer()
+	b.ReportAllocs()
+
+	for i := 0; i < b.N; i++ {
+		for _, value := range intValues {
+			if err := dvr.Format(ctx, value, buf); err != nil {
 				b.Fatal(err)
 			}
 		}

--- a/tx/textual/valuerenderer/dec.go
+++ b/tx/textual/valuerenderer/dec.go
@@ -34,13 +34,13 @@ func (vr decValueRenderer) Parse(_ context.Context, r io.Reader) (protoreflect.V
 // object).
 func formatDecimal(v string) (string, error) {
 	parts := strings.Split(v, ".")
+	if len(parts) > 2 {
+		return "", fmt.Errorf("invalid decimal: too many points in %s", v)
+	}
+
 	intPart, err := formatInteger(parts[0])
 	if err != nil {
 		return "", err
-	}
-
-	if len(parts) > 2 {
-		return "", fmt.Errorf("invalid decimal %s", v)
 	}
 
 	if len(parts) == 1 {
@@ -50,6 +50,12 @@ func formatDecimal(v string) (string, error) {
 	decPart := strings.TrimRight(parts[1], "0")
 	if len(decPart) == 0 {
 		return intPart, nil
+	}
+
+	// Ensure that the decimal part has only digits.
+	// https://github.com/cosmos/cosmos-sdk/issues/12811
+	if !hasOnlyDigits(decPart) {
+		return "", fmt.Errorf("non-digits detected after decimal point in: %q", decPart)
 	}
 
 	return intPart + "." + decPart, nil

--- a/tx/textual/valuerenderer/dec_test.go
+++ b/tx/textual/valuerenderer/dec_test.go
@@ -1,0 +1,35 @@
+package valuerenderer
+
+import (
+	"strings"
+	"testing"
+)
+
+func TestFormatDecimalNonDigits(t *testing.T) {
+	badCases := []string{
+		"10.a",
+		"1a.10",
+		"p1a10.",
+		"0.10p",
+		"--10",
+		"12.😎😎",
+		"11111111111133333333333333333333333333333a",
+		"11111111111133333333333333333333333333333 192892",
+	}
+
+	for _, value := range badCases {
+		value := value
+		t.Run(value, func(t *testing.T) {
+			s, err := formatDecimal(value)
+			if err == nil {
+				t.Fatal("Expected an error")
+			}
+			if g, w := err.Error(), "non-digits"; !strings.Contains(g, w) {
+				t.Errorf("Error mismatch\nGot:  %q\nWant substring: %q", g, w)
+			}
+			if s != "" {
+				t.Fatalf("Got a non-empty string: %q", s)
+			}
+		})
+	}
+}

--- a/tx/textual/valuerenderer/int.go
+++ b/tx/textual/valuerenderer/int.go
@@ -2,6 +2,7 @@ package valuerenderer
 
 import (
 	"context"
+	"fmt"
 	"io"
 	"strings"
 
@@ -26,6 +27,18 @@ func (vr intValueRenderer) Parse(_ context.Context, r io.Reader) (protoreflect.V
 	panic("implement me")
 }
 
+func hasOnlyDigits(s string) bool {
+	if s == "" {
+		return false
+	}
+	for _, r := range s {
+		if r < '0' || r > '9' {
+			return false
+		}
+	}
+	return true
+}
+
 // formatInteger formats an integer into a value-rendered string. This function
 // operates with string manipulation (instead of manipulating the int or sdk.Int
 // object).
@@ -37,7 +50,11 @@ func formatInteger(v string) (string, error) {
 	}
 	if len(v) > 1 {
 		v = strings.TrimLeft(v, "0")
+	}
 
+	// Ensure that the string contains only digits at this point.
+	if !hasOnlyDigits(v) {
+		return "", fmt.Errorf("expecting only digits 0-9, but got non-digits in %q", v)
 	}
 
 	startOffset := 3

--- a/tx/textual/valuerenderer/int_test.go
+++ b/tx/textual/valuerenderer/int_test.go
@@ -1,0 +1,35 @@
+package valuerenderer
+
+import (
+	"strings"
+	"testing"
+)
+
+func TestFormatIntegerNonDigits(t *testing.T) {
+	badCases := []string{
+		"a10",
+		"1a10",
+		"p1a10",
+		"10p",
+		"--10",
+		"😎😎",
+		"11111111111133333333333333333333333333333a",
+		"11111111111133333333333333333333333333333 192892",
+	}
+
+	for _, value := range badCases {
+		value := value
+		t.Run(value, func(t *testing.T) {
+			s, err := formatInteger(value)
+			if err == nil {
+				t.Fatal("Expected an error")
+			}
+			if g, w := err.Error(), "but got non-digits in"; !strings.Contains(g, w) {
+				t.Errorf("Error mismatch\nGot:  %q\nWant substring: %q", g, w)
+			}
+			if s != "" {
+				t.Fatalf("Got a non-empty string: %q", s)
+			}
+		})
+	}
+}

